src/wx/panel.cpp

```diff
@@ -973,6 +961,307 @@ void GameArea::Resume()
     SetFocus();
 }
 
+IMPLEMENT_DYNAMIC_CLASS(GameArea, wxPanel)
+
+GameArea::GameArea()
+    : wxPanel()
+    , panel(NULL)
+    , emusys(NULL)
+    , was_paused(false)
+    , rewind_time(0)
+    , do_rewind(false)
+    , rewind_mem(0)
+    , num_rewind_states(0)
+    , loaded(IMAGE_UNKNOWN)
+    , basic_width(GBAWidth)
+    , basic_height(GBAHeight)
+    , fullscreen(false)
+    , paused(false)
+    , pointer_blanked(false)
+    , mouse_active_time(0)
+{
+    SetSizer(new wxBoxSizer(wxVERTICAL));
+    // all renderers prefer 32-bit
+    // well, "simple" prefers 24-bit, but that's not available for filters
+    systemColorDepth = 32;
+    hq2x_init(32);
+    Init_2xSaI(32);
+
+#ifndef NO_THREAD_MAINLOOP
+    Bind(WX_THREAD_REQUEST_UPDATEDRAWPANEL, &GameArea::RequestUpdateDrawPanel, this);
+    Bind(WX_THREAD_REQUEST_DRAWFRAME, &GameArea::RequestDrawFrame, this);
+    Bind(WX_THREAD_REQUEST_UPDATESTATUSBAR, &GameArea::RequestUpdateStatusBar, this);
+    Bind(WX_THREAD_REQUEST_GBPRINTER, &GameArea::ShowPrinter, this);
+    Bind(WX_THREAD_REQUEST_UPDATELOG, &GameArea::RequestUpdateLog, this);
+#endif
+}
+
+#ifndef NO_THREAD_MAINLOOP
+wxCriticalSection MainFrame::emulationCS;
+
+void GameArea::RequestDraw()
+{
+    wxQueueEvent(this, new wxThreadEvent(WX_THREAD_REQUEST_DRAWFRAME));
+}
+
+void GameArea::RequestDrawFrame(wxThreadEvent& WXUNUSED(event))
+{
+    wxCriticalSectionLocker lock(MainFrame::emulationCS);
+    MainFrame* mf = wxGetApp().frame;
+    mf->UpdateViewers();
+    if (panel) {
+        panel->DrawArea(&pix);
+    }
+}
+
+wxThread::ExitCode GameArea::Entry()
+{
+    while (!GetThread()->TestDestroy()) {
+        {
+            wxCriticalSectionLocker lock(MainFrame::emulationCS);
+            if (emusys) {
+                wxGetApp().frame->PollJoysticks();
+                if (!panel) {
+                    wxQueueEvent(this, new wxThreadEvent(WX_THREAD_REQUEST_UPDATEDRAWPANEL));
+                }
+                if (!paused && panel) {
+                    emusys->emuMain(emusys->emuCount);
+                }
+	    }
+        }
+        wxMilliSleep(1);
```

Comment:
@rkitover Here we can play a little bit about how to do proper timesteps and other controls. We can try to control when and for how long we should sleep, like an actual main loop for a game.